### ISSUE TITLE
TypeError: Cannot read property 'render'of undefined

### ISSUE THREAD

**Issue by cjduncana** [cmstead/express-route-fake#1]
# TL;DR

I want to make sure that the client receives a 200 HTTP status code when they visit the index page. When I run the test, a TypeError appears. How can I use render in my routes and still unit test?
## Problem

This package seems perfect for unit testing, but I believe that I am not using it as supposed.
### controllers.js

``` javascript
var router = require('express').Router();

router.get('/', function(request, response) {
    response.status(200).render('index');
});

module.exports = router;
```

Simple enough! A router that when it receives a request for the index page, it renders a template for the index page and responds with the rendered page and a 200 HTTP status code.
### test.js The Beginning

``` javascript
var should = require('should');
var mockery = require('mockery');
var expressRouteFake = require('express-route-fake');

describe('Testing the Index Page Controllers:', function() {
    var controllers;
    var req;
    var res;

    beforeEach(function() {
        req = {};
        res = {
            resData: {
                status: 0,
                response: ''
            },
            status: function(status) {
                res.resData.status = status;
            },
            send: function(response) {
                res.resData.response = response;
            },
            end: function() {}
        };
        mockery.resetCache();
        mockery.enable({
            warnOnReplace: false,
            warnOnUnregistered: false
        });
        expressRouteFake.reset();
        mockery.registerMock('express', expressRouteFake);
        controllers = require('./controllers');
    });
```

This part is almost a carbon-copy from the code in [your article](http://www.chrisstead.com/archives/1128/unit-testing-express-routing/). The changes that I made are just two:
1.   Rename `expressFake` for `expressRouteFake`.
2.   Load `controllers` as the module being tested.
### test.js The Test

``` javascript
    it('Get the Index Page', function(done) {
        var routeAction = expressRouteFake.getRouteAction('get', '/');
        routeAction(req, res);
        res.resData.status.should.be.equal(200);
        return done();
    });
```

Again, this is also similar to [your article](http://www.chrisstead.com/archives/1128/unit-testing-express-routing/), except that I use the `should` library for assertion.
### test.js The End

``` javascript
    afterEach(function() {
        mockery.deregisterAll();
        mockery.disable();
    });
});
```

This part I did not change at all.
### Run `npm test`

I run the test, but I get the following error:

```
1) Testing the Index Page Controllers: Get the Index Page:
     TypeError: Cannot read property 'render' of undefined
      at controllers.js:6:25
      at Context.<anonymous> (test.js:64:9)
```

I do not know why, and I do not understand why either. Could you help me with this?

**Comment by cmstead** [comment]
The root of the problem is that I didn't build a render function in, but I definitely could with minimal effort. Sorry for the confusion, I'll try to make a little time this evening to add it for you.

**Comment by cjduncana** [comment]
If there is anything I can do to help, just let me know.

**Comment by cmstead** [comment]
I'll gladly accept pull requests if you already have a solution which works
for you. :-)

**Comment by cjduncana** [comment]
Let me try to code this. Any direction would be appreciated. I am still trying to wrap my head around the bind function. Also, I do not understand what exactly happens once I initialize the function `routeAction`, and then use it in `routeAction(req, res);`.

**Comment by cjduncana** [comment]
Now, I feel that I have a better grasp around `bind()`, but I still do not understand how it is used in your module, and how does that relate to the rest of your module.

**Comment by cjduncana** [comment]
I'm now trying to understand why when I call `var routeAction = expressRouteFake.getRouteAction('get', '/');`
it does not throw an `Express Route Fake Error`? As far as I can tell, at the moment this is called, `routeFunctions['get'] == {} != { '/': {} }`, unless when `mockery.registerMock()` was called this was changed. My investigation continues.

**Comment by cmstead** [comment]
I updated the core fake with an app.render function, however, after reviewing your code again, the response object you are creating with your mock is missing a render function, and your send needs to return the response object so render is available.

```
res = {
        resData: {
            status: 0,
            response: ''
        },
        status: function(status) {
            res.resData.status = status;
        },
        send: function(response) {
            res.resData.response = response;
            return res;
        },
        render: function () {
            // add any logic you are interested in here.
        },
        end: function() {}
    };
```

I will add an issue to make generating request and response fakes easier so this kind of issue is easier to avoid.

**Comment by cmstead** [comment]
Sorry, I'm catching up with your comments.  So, the bind function is being called to bind the HTTP verbs to the routeStub function so it can correctly capture the verb which each route belongs to.

routeStub captures the path to your endpoint and the action it should perform when someone makes a call to your route so you can call the function directly and test that the internal logic is correct. There is a lot of pointer management happening. The complexity of managing all of the pointers to the route actions is why I built this originally. Testing express routes is tough, so this basically gives you direct access to your API functions. : )

Let me know if this helps. I might write a second blog post to help make more sense of what is happening in this module.

**Comment by cjduncana** [comment]
I still get the `TypeError` for `render` for

``` javascript
render: function(render) {
    res.resData.response = render;
    return res;
}
```

and

``` javascript
render: function() {
    return 'It rendered!';
}
```

**Comment by cjduncana** [comment]
BTW, looking forward for your next article.
